Add long error explanation for E0724 (#73163)
rust-lang/rust: src/librustc_error_codes/error_codes/E0724.md

`#[ffi_returns_twice]` was used on non-foreign function.

Erroneous code example:

```compile_fail,E0724
#![feature(ffi_returns_twice)]
#![crate_type = "lib"]

#[ffi_returns_twice] // error!
pub fn foo() {}
```

> `#[ffi_returns_twice]` can only be used on foreign functions.

Reviewer: ```suggestion
`#[ffi_returns_twice]` can only be used on foreign function declarations. 
```

> For example, we might resolve the previous example by declaring

Reviewer: ```suggestion
For example, we might correct the previous example by declaring
```

> inside an `extern` block.

Reviewer: ```suggestion
the function inside of an `extern` block.
```

`#[ffi_returns_twice]` was used on non-foreign function.

Erroneous code example:

```compile_fail,E0724
#![feature(ffi_returns_twice)]
#![crate_type = "lib"]

#[ffi_returns_twice] // error!
pub fn foo() {}
```

`#[ffi_returns_twice]` can only be used on foreign functions.
For example, we might resolve the previous example by declaring

> inside an `extern` block annotated with a `#[link]` attribute containing

Reviewer: You’ll need to update this sentence to reflect the changed example that follows.

`#[ffi_returns_twice]` was used on non-foreign function.

Erroneous code example:

```compile_fail,E0724
#![feature(ffi_returns_twice)]
#![crate_type = "lib"]

#[ffi_returns_twice] // error!
pub fn foo() {}
```

`#[ffi_returns_twice]` can only be used on foreign functions.
For example, we might resolve the previous example by declaring
inside an `extern` block annotated with a `#[link]` attribute containing
the name of the foreign library.

```
#![feature(ffi_returns_twice)]
#![crate_type = "lib"]

#[link(name = "m")]

> extern {

Reviewer: I think that this example would be stronger if it didn't also introduce the `#[link]` attribute - if a user doesn't know what this is, then it could just add confusion - an `extern "C" { }` block is probably as simple as we could do.

```rust
extern "C" {
    #[ffi_returns_twice] // ok!
    pub fn foo();
}
```
